From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Andrew Youll <adyoull@users.noreply.github.com>
Date: Sat, 26 Sep 2026 22:47:54 +0100
Subject: [PATCH 10/11] RISC OS: unaligned-safe loads and stores in the script
 save code

Script state in save games packs UDWORD/SWORD/... values of mixed sizes
into a byte buffer (eventsave.c, scriptobj.c), so most of them are at
unaligned addresses; RISC OS traps unaligned accesses. They now use
endian_hack.h's wz_load_* / wz_store_* (the two function pointers are
copied with memcpy). Found with -Wcast-align=strict
(tools/check-unaligned.sh in riscos-warzone2100); the save format doesn't
change.
---
 lib/script/eventsave.c | 75 ++++++++++++++++++++++--------------------
 src/scriptobj.c        | 52 ++++++++++++++---------------
 2 files changed, 65 insertions(+), 62 deletions(-)

diff --git a/lib/script/eventsave.c b/lib/script/eventsave.c
index f28e804..a1fc0e7 100644
--- a/lib/script/eventsave.c
+++ b/lib/script/eventsave.c
@@ -54,7 +54,7 @@ static BOOL eventSaveContext(char *pBuffer, UDWORD *pSize)
 	char				*pPos;
 //not hashed	char				*pScriptID;
 	UDWORD				hashedName;
-	UWORD				*pValSize = NULL;
+	char				*pValSize = NULL;
 
 
 	size = 0;
@@ -87,11 +87,11 @@ static BOOL eventSaveContext(char *pBuffer, UDWORD *pSize)
 		{
 //not hashed			strcpy(pPos, pScriptID);
 //not hashed			pPos += strlen(pScriptID) + 1;
-			*((UDWORD*)pPos) = (UDWORD)hashedName;
+			wz_store_u32(pPos, (UDWORD)hashedName);
 			endian_udword((UDWORD*)pPos);
 			pPos += sizeof(UDWORD);
 
-			*((SWORD*)pPos) = (SWORD)numVars;
+			wz_store_s16(pPos, (SWORD)numVars);
 			endian_sword((SWORD*)pPos);
 			pPos += sizeof(SWORD);
 
@@ -115,7 +115,7 @@ static BOOL eventSaveContext(char *pBuffer, UDWORD *pSize)
 					ASSERT( psVal->type < SWORD_MAX,
 						"eventSaveContext: variable type number too big" );
 
-					*((SWORD*)pPos) = (SWORD)psVal->type;
+					wz_store_s16(pPos, (SWORD)psVal->type);
 					endian_sword((SWORD*)pPos);
 
 					pPos += sizeof(SWORD);
@@ -134,7 +134,7 @@ static BOOL eventSaveContext(char *pBuffer, UDWORD *pSize)
 
 					if (pBuffer != NULL)
 					{
-						*((UDWORD *)pPos) = stringLen;
+						wz_store_u32(pPos, stringLen);
 						endian_udword((UDWORD *)pPos);
 						pPos += sizeof(UDWORD);
 
@@ -153,7 +153,7 @@ static BOOL eventSaveContext(char *pBuffer, UDWORD *pSize)
 					if (pBuffer != NULL)
 					{
 /* FIXME: this does not work for VAL_OBJ_GETSET, VAL_FUNC_EXTERN */
-						*((UDWORD *)pPos) = (UDWORD)psVal->v.ival;
+						wz_store_u32(pPos, (UDWORD)psVal->v.ival);
 						endian_udword((UDWORD*)pPos);
 						
 						pPos += sizeof(UDWORD);
@@ -172,7 +172,7 @@ static BOOL eventSaveContext(char *pBuffer, UDWORD *pSize)
 					// reserve some space to store how many bytes the value uses
 					if (pBuffer != NULL)
 					{
-						pValSize = (UWORD *)pPos;
+						pValSize = pPos;
 						pPos += sizeof(UWORD);
 					}
 					size += sizeof(UWORD);
@@ -186,8 +186,11 @@ static BOOL eventSaveContext(char *pBuffer, UDWORD *pSize)
 
 					if (pBuffer != NULL)
 					{
-						*pValSize = (UWORD)valSize;
-						endian_uword((UWORD*)pValSize);
+						{
+							UWORD wz_vs = (UWORD)valSize;
+							endian_uword(&wz_vs);
+							wz_store_u16(pValSize, wz_vs);
+						}
 
 						pPos += valSize;
 					}
@@ -211,7 +214,7 @@ static BOOL eventSaveContext(char *pBuffer, UDWORD *pSize)
 	// actually store how many contexts have been saved
 	if (pBuffer != NULL)
 	{
-		*((SWORD *)pBuffer) = (SWORD)numContext;
+		wz_store_s16(pBuffer, (SWORD)numContext);
 		endian_sword((SWORD*)pBuffer);
 	}
 	*pSize = size;
@@ -239,7 +242,7 @@ static BOOL eventLoadContext(const SDWORD version, char *pBuffer, UDWORD *pSize,
 
 	// get the number of contexts in the save file
 	endian_sword((SWORD*)pPos);
-	numContext = *((SWORD *)pPos);
+	numContext = wz_load_s16(pPos);
 	pPos += sizeof(SWORD);
 	size += sizeof(SWORD);
 
@@ -248,7 +251,7 @@ static BOOL eventLoadContext(const SDWORD version, char *pBuffer, UDWORD *pSize,
 	{
 	    if(bHashed) {
     		endian_udword((UDWORD*)pPos);
-    		hashedName = *((UDWORD*)pPos);
+    		hashedName = wz_load_u32(pPos);
     		psCode = (SCRIPT_CODE*)resGetDataFromHash("SCRIPT", hashedName);
     		pPos += sizeof(UDWORD);
 	    } else {
@@ -261,10 +264,10 @@ static BOOL eventLoadContext(const SDWORD version, char *pBuffer, UDWORD *pSize,
 		endian_sword((SWORD*)pPos);
    		numVars = psCode->numGlobals + psCode->arraySize;
         
-		if (numVars != *((SWORD*)pPos))
+		if (numVars != wz_load_s16(pPos))
 		{
 			ASSERT(false, "Context %d of %d: Number of context variables (%d) does not match the script code (%d)", 
-			       context, numContext, numVars, *((SWORD*)pPos));
+			       context, numContext, numVars, wz_load_s16(pPos));
 			return false;
 		}
 		pPos += sizeof(SWORD);
@@ -292,7 +295,7 @@ static BOOL eventLoadContext(const SDWORD version, char *pBuffer, UDWORD *pSize,
 		{
 			// get the variable type
 			endian_sword((SWORD*)pPos);
-			type = (INTERP_TYPE) *((SWORD*)pPos);
+			type = (INTERP_TYPE) wz_load_s16(pPos);
 			pPos += sizeof(SWORD);
 			size += sizeof(SWORD);
 
@@ -305,12 +308,12 @@ static BOOL eventLoadContext(const SDWORD version, char *pBuffer, UDWORD *pSize,
 
 				switch (type) {
 				case VAL_BOOL:
-					data.v.bval = *((BOOL*)pPos);
+					data.v.bval = wz_load_s32(pPos);
 					pPos += sizeof(BOOL);
 					size += sizeof(BOOL);
 					break;
 				case VAL_FLOAT:
-					data.v.fval = *((float*)pPos);
+					data.v.fval = wz_load_float(pPos);
 					pPos += sizeof(float);
 					size += sizeof(float);
 					break;
@@ -320,7 +323,7 @@ static BOOL eventLoadContext(const SDWORD version, char *pBuffer, UDWORD *pSize,
 				case VAL_VOID:
 				case VAL_OPCODE:
 				case VAL_PKOPCODE:
-					data.v.ival = *((UDWORD *)pPos);
+					data.v.ival = wz_load_u32(pPos);
 					pPos += sizeof(UDWORD);
 					size += sizeof(UDWORD);
 					break;
@@ -328,7 +331,7 @@ static BOOL eventLoadContext(const SDWORD version, char *pBuffer, UDWORD *pSize,
 					data.v.sval = (char*)malloc(MAXSTRLEN);
 					strcpy(data.v.sval, "\0");
 
-					stringLen = *((UDWORD *)pPos);	//read string length
+					stringLen = wz_load_u32(pPos);	//read string length
 					
 					pPos += sizeof(UDWORD);
 					size += sizeof(UDWORD);
@@ -343,13 +346,13 @@ static BOOL eventLoadContext(const SDWORD version, char *pBuffer, UDWORD *pSize,
 					break;
 				case VAL_OBJ_GETSET:
 /* FIXME: saving pointer on disk! */
-					data.v.pObjGetSet = *((SCRIPT_VARFUNC*)pPos);
+					memcpy(&data.v.pObjGetSet, pPos, sizeof(SCRIPT_VARFUNC));	// may be unaligned
 					pPos += sizeof(SCRIPT_VARFUNC);
 					size += sizeof(SCRIPT_VARFUNC);
 					break;
 				case VAL_FUNC_EXTERN:
 /* FIXME: saving pointer on disk! */
-					data.v.pFuncExtern = *((SCRIPT_FUNC*)pPos);
+					memcpy(&data.v.pFuncExtern, pPos, sizeof(SCRIPT_FUNC));	// may be unaligned
 					pPos += sizeof(SCRIPT_FUNC);
 					size += sizeof(SCRIPT_FUNC);
 					break;
@@ -374,7 +377,7 @@ static BOOL eventLoadContext(const SDWORD version, char *pBuffer, UDWORD *pSize,
 					"eventLoadContext: no load function for type %d\n", type );
 
 				endian_uword((UWORD*)pPos);
-				valSize = *((UWORD *)pPos);
+				valSize = wz_load_u16(pPos);
 
 				pPos += sizeof(UWORD);
 				size += sizeof(UWORD);
@@ -468,7 +471,7 @@ static BOOL eventSaveTriggerList(ACTIVE_TRIGGER *psList, char *pBuffer, UDWORD *
 
 		if (pBuffer != NULL)
 		{
-			*((UDWORD*)pPos) = psCurr->testTime;
+			wz_store_u32(pPos, psCurr->testTime);
 			endian_udword((UDWORD*)pPos);
 
 			pPos += sizeof(UDWORD);
@@ -478,19 +481,19 @@ static BOOL eventSaveTriggerList(ACTIVE_TRIGGER *psList, char *pBuffer, UDWORD *
 				abort();
 				return false;
 			}
-			*((SWORD*)pPos) = (SWORD)context;
+			wz_store_s16(pPos, (SWORD)context);
 			endian_sword((SWORD*)pPos);
 			pPos += sizeof(SWORD);
-			*((SWORD*)pPos) = psCurr->type;
+			wz_store_s16(pPos, psCurr->type);
 			endian_sword((SWORD*)pPos);
 			pPos += sizeof(SWORD);
-			*((SWORD*)pPos) = psCurr->trigger;
+			wz_store_s16(pPos, psCurr->trigger);
 			endian_sword((SWORD*)pPos);
 			pPos += sizeof(SWORD);
-			*((UWORD*)pPos) = psCurr->event;
+			wz_store_u16(pPos, psCurr->event);
 			endian_uword((UWORD*)pPos);
 			pPos += sizeof(UWORD);
-			*((UWORD*)pPos) = psCurr->offset;
+			wz_store_u16(pPos, psCurr->offset);
 			endian_uword((UWORD*)pPos);
 			pPos += sizeof(UWORD);
 		}
@@ -498,7 +501,7 @@ static BOOL eventSaveTriggerList(ACTIVE_TRIGGER *psList, char *pBuffer, UDWORD *
 	}
 	if (pBuffer != NULL)
 	{
-		*((SDWORD*)pBuffer) = numTriggers;
+		wz_store_s32(pBuffer, numTriggers);
 		endian_sdword((SDWORD*)pBuffer);
 	}
 
@@ -521,18 +524,18 @@ static BOOL eventLoadTriggerList(WZ_DECL_UNUSED const SDWORD version, char *pBuf
 
 	// get the number of triggers
 	endian_sdword((SDWORD*)pPos);
-	numTriggers = *((SDWORD*)pPos);
+	numTriggers = wz_load_s32(pPos);
 	pPos += sizeof(SDWORD);
 	size += sizeof(SDWORD);
 
 	for(i=0; i<numTriggers; i+= 1)
 	{
 		endian_udword((UDWORD*)pPos);
-		time = *((UDWORD*)pPos);
+		time = wz_load_u32(pPos);
 		pPos += sizeof(UDWORD);
 
 		endian_sword((SWORD*)pPos);
-		context = *((SWORD*)pPos);
+		context = wz_load_s16(pPos);
 		pPos += sizeof(SWORD);
 		if (!eventFindContext(context, &psContext))
 		{
@@ -542,19 +545,19 @@ static BOOL eventLoadTriggerList(WZ_DECL_UNUSED const SDWORD version, char *pBuf
 		}
 
 		endian_sword((SWORD*)pPos);
-		type = *((SWORD*)pPos);
+		type = wz_load_s16(pPos);
 		pPos += sizeof(SWORD);
 
 		endian_sword((SWORD*)pPos);
-		trigger = *((SWORD*)pPos);
+		trigger = wz_load_s16(pPos);
 		pPos += sizeof(SWORD);
 
 		endian_uword((UWORD*)pPos);
-		event = *((UWORD*)pPos);
+		event = wz_load_u16(pPos);
 		pPos += sizeof(UWORD);
 
 		endian_uword((UWORD*)pPos);
-		offset = *((UWORD*)pPos);
+		offset = wz_load_u16(pPos);
 		pPos += sizeof(UWORD);
 
 		size += sizeof(UDWORD) + sizeof(SWORD)*3 + sizeof(UWORD)*2;
diff --git a/src/scriptobj.c b/src/scriptobj.c
index 612c81c..a914047 100644
--- a/src/scriptobj.c
+++ b/src/scriptobj.c
@@ -772,11 +772,11 @@ BOOL scrValDefSave(INTERP_VAL *psVal, char *pBuffer, UDWORD *pSize)
 		{
 			if (psVal->v.oval == NULL || ((BASE_OBJECT *)psVal->v.oval)->died > 1 || !getBaseObjFromId(((BASE_OBJECT *)psVal->v.oval)->id))
 			{
-				*((UDWORD*)pBuffer) = UDWORD_MAX;
+				wz_store_u32(pBuffer, UDWORD_MAX);
 			}
 			else
 			{
-				*((UDWORD*)pBuffer) = ((BASE_OBJECT *)psVal->v.oval)->id;
+				wz_store_u32(pBuffer, ((BASE_OBJECT *)psVal->v.oval)->id);
 			}
 			endian_udword((UDWORD*)pBuffer);
 		}
@@ -813,11 +813,11 @@ BOOL scrValDefSave(INTERP_VAL *psVal, char *pBuffer, UDWORD *pSize)
 		{
 			if (psVal->v.oval == NULL)
 			{
-				*((UDWORD*)pBuffer) = UDWORD_MAX;
+				wz_store_u32(pBuffer, UDWORD_MAX);
 			}
 			else
 			{
-				*((UDWORD*)pBuffer) = ((DROID_TEMPLATE *)psVal->v.oval)->multiPlayerID;
+				wz_store_u32(pBuffer, ((DROID_TEMPLATE *)psVal->v.oval)->multiPlayerID);
 			}
 			endian_udword((UDWORD*)pBuffer);
 		}
@@ -830,7 +830,7 @@ BOOL scrValDefSave(INTERP_VAL *psVal, char *pBuffer, UDWORD *pSize)
 
 		if (pBuffer)
 		{
-			*((uint16_t*)pBuffer) = len;
+			wz_store_u16(pBuffer, len);
 			endian_uword((uint16_t*)pBuffer);
 
 			memcpy(pBuffer + sizeof(len), idStr, len);
@@ -884,26 +884,26 @@ BOOL scrValDefSave(INTERP_VAL *psVal, char *pBuffer, UDWORD *pSize)
 		{
 			pPos = pBuffer;
 
-			*((SDWORD *)pPos) = members;
+			wz_store_s32(pPos, members);
 			endian_sdword((SDWORD*)pPos);
 			pPos += sizeof(SDWORD);
 
 			if (psGroup)
 			{
 				// store the run data
-				*((SDWORD *)pPos) = psGroup->sRunData.sPos.x;
+				wz_store_s32(pPos, psGroup->sRunData.sPos.x);
 				endian_sdword((SDWORD*)pPos);
 				pPos += sizeof(SDWORD);
-				*((SDWORD *)pPos) = psGroup->sRunData.sPos.y;
+				wz_store_s32(pPos, psGroup->sRunData.sPos.y);
 				endian_sdword((SDWORD*)pPos);
 				pPos += sizeof(SDWORD);
-				*((SDWORD *)pPos) = psGroup->sRunData.forceLevel;
+				wz_store_s32(pPos, psGroup->sRunData.forceLevel);
 				endian_sdword((SDWORD*)pPos);
 				pPos += sizeof(SDWORD);
-				*((SDWORD *)pPos) = psGroup->sRunData.leadership;
+				wz_store_s32(pPos, psGroup->sRunData.leadership);
 				endian_sdword((SDWORD*)pPos);
 				pPos += sizeof(SDWORD);
-				*((SDWORD *)pPos) = psGroup->sRunData.healthLevel;
+				wz_store_s32(pPos, psGroup->sRunData.healthLevel);
 				endian_sdword((SDWORD*)pPos);
 				pPos += sizeof(SDWORD);
 
@@ -912,7 +912,7 @@ BOOL scrValDefSave(INTERP_VAL *psVal, char *pBuffer, UDWORD *pSize)
 				{
 					checkValidId(psCDroid->id);
 
-					*((UDWORD *)pPos) = psCDroid->id;
+					wz_store_u32(pPos, psCDroid->id);
 					endian_udword((UDWORD*)pPos);
 
 					pPos += sizeof(UDWORD);
@@ -957,7 +957,7 @@ BOOL scrValDefSave(INTERP_VAL *psVal, char *pBuffer, UDWORD *pSize)
 		// just save the variable contents directly
 		if (pBuffer)
 		{
-			*((UDWORD *)pBuffer) = psVal->v.ival;
+			wz_store_u32(pBuffer, psVal->v.ival);
 			endian_udword((UDWORD*)pBuffer);
 		}
 		*pSize = sizeof(UDWORD);
@@ -1002,7 +1002,7 @@ BOOL scrValDefLoad(SDWORD version, INTERP_VAL *psVal, char *pBuffer, UDWORD size
 	case ST_DROID:
 	case ST_STRUCTURE:
 	case ST_FEATURE:
-		id = *((UDWORD *)pBuffer);
+		id = wz_load_u32(pBuffer);
 		endian_udword(&id);
 
 		if (id == UDWORD_MAX)
@@ -1123,7 +1123,7 @@ BOOL scrValDefLoad(SDWORD version, INTERP_VAL *psVal, char *pBuffer, UDWORD size
 		psVal->v.ival = index;
 		break;
 	case ST_TEMPLATE:
-		id = *((UDWORD *)pBuffer);
+		id = wz_load_u32(pBuffer);
 		endian_udword(&id);
 
 		if (id == UDWORD_MAX)
@@ -1149,7 +1149,7 @@ BOOL scrValDefLoad(SDWORD version, INTERP_VAL *psVal, char *pBuffer, UDWORD size
 				return false;
 			}
 
-			id = *((UDWORD*)pBuffer);
+			id = wz_load_u32(pBuffer);
 			endian_udword(&id);
 
 			if (id == UDWORD_MAX)
@@ -1180,7 +1180,7 @@ BOOL scrValDefLoad(SDWORD version, INTERP_VAL *psVal, char *pBuffer, UDWORD size
 				return false;
 			}
 
-			len = *((uint16_t*)pBuffer);
+			len = wz_load_u16(pBuffer);
 			endian_uword(&len);
 
 			if (size < sizeof(len) + len)
@@ -1292,11 +1292,11 @@ BOOL scrValDefLoad(SDWORD version, INTERP_VAL *psVal, char *pBuffer, UDWORD size
 
 			// get saved group member count/nullpointer flag
 			endian_sdword((SDWORD*)pPos);
-			bObjectDefined = ( *((SDWORD *)pPos) != UNALLOCATED_OBJECT );
+			bObjectDefined = ( wz_load_s32(pPos) != UNALLOCATED_OBJECT );
 
 			if(bObjectDefined)
 			{
-				savedMembers = *((SDWORD *)pPos);	// get number of saved group members
+				savedMembers = wz_load_s32(pPos);	// get number of saved group members
 
 				ASSERT(savedMembers == members, "scrValDefLoad: calculated and saved group member count did not match." );
 			}
@@ -1311,22 +1311,22 @@ BOOL scrValDefLoad(SDWORD version, INTERP_VAL *psVal, char *pBuffer, UDWORD size
 				// load the retreat data
 				psGroup = (DROID_GROUP*)(psVal->v.oval);
 				endian_sdword((SDWORD*)pPos);
-				psGroup->sRunData.sPos.x = *((SDWORD *)pPos);
+				psGroup->sRunData.sPos.x = wz_load_s32(pPos);
 				pPos += sizeof(SDWORD);
 				endian_sdword((SDWORD*)pPos);
-				psGroup->sRunData.sPos.y = *((SDWORD *)pPos);
+				psGroup->sRunData.sPos.y = wz_load_s32(pPos);
 				pPos += sizeof(SDWORD);
 				endian_sdword((SDWORD*)pPos);
-				psGroup->sRunData.forceLevel = (UBYTE)(*((SDWORD *)pPos));
+				psGroup->sRunData.forceLevel = (UBYTE)(wz_load_s32(pPos));
 				pPos += sizeof(SDWORD);
 				endian_sdword((SDWORD*)pPos);
-				psGroup->sRunData.leadership = (UBYTE)(*((SDWORD *)pPos));
+				psGroup->sRunData.leadership = (UBYTE)(wz_load_s32(pPos));
 				pPos += sizeof(SDWORD);
 			}
 			if (version >= 3)
 			{
 				endian_sdword((SDWORD*)pPos);
-				psGroup->sRunData.healthLevel = (UBYTE)(*((SDWORD *)pPos));
+				psGroup->sRunData.healthLevel = (UBYTE)(wz_load_s32(pPos));
 				pPos += sizeof(SDWORD);
 			}
 
@@ -1334,7 +1334,7 @@ BOOL scrValDefLoad(SDWORD version, INTERP_VAL *psVal, char *pBuffer, UDWORD size
 			while (members > 0)
 			{
 				endian_udword((UDWORD*)pPos);
-				id = *((UDWORD *) pPos);
+				id = wz_load_u32(pPos);
 				psCDroid = (DROID *)getBaseObjFromId(id);
 				if (!psCDroid)
 				{
@@ -1382,7 +1382,7 @@ BOOL scrValDefLoad(SDWORD version, INTERP_VAL *psVal, char *pBuffer, UDWORD size
 	case ST_DROIDID:
 	default:
 		// just set the contents directly
-		psVal->v.ival = *((SDWORD *)pBuffer);
+		psVal->v.ival = wz_load_s32(pBuffer);
 		endian_sdword(&psVal->v.ival);
 		break;
 	}
